### PR TITLE
 fix(config): restore custom logger feature 

### PR DESCRIPTION
**Pull request by Qard**
This PR restores the custom logger feature.
Closes elastic/apm-agent-nodejs#289

### CODE REVIEW

File: lib/agent.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ var parseUrl = require('url').parse
 var path = require('path')
 
 var afterAll = require('after-all-results')
-var debug = require('debug')('elastic-apm')
+var ConsoleLogLevel = require('console-log-level')
```
Comment (watson) [inline lib/agent.js]: It's kind of a convention that variables that start with an uppercase letter like this needs to be instantiated with the `new` keyword, so I'd probably change it to `consoleLogLevel`

File: lib/agent.js [hunk 2]
```diff
@@ -68,7 +67,11 @@ Agent.prototype._config = function (opts) {
     ? parseUrl(this._conf.serverUrl).hostname
     : 'localhost'
 
-  logger.init({level: this._conf.logLevel})
+  this.logger = this._conf.logger || ConsoleLogLevel
+
+  if (typeof this.logger === 'function') {
+    this.logger = this.logger({ level: this._conf.logLevel })
```
Comment (watson) [inline lib/agent.js]: Is it safe to assume that all loggers that are a function will have this API?

File: lib/parsers.js [hunk 3]
```diff
@@ -39,7 +38,8 @@ exports.parseMessage = function (msg) {
 exports.parseError = function (err, agent, cb) {
   stackman.callsites(err, function (_err, callsites) {
     if (_err) {
-      debug('error while getting error callsites: %s', _err.message)
+      console.log(agent)
```
Comment (watson) [inline lib/parsers.js]: Ups 😄

File: lib/instrumentation/shimmer.js [hunk 4]
```diff
@@ -15,7 +15,7 @@
  * BSD-2-Clause, http://opensource.org/licenses/BSD-2-Clause
  */
 
-var debug = require('debug')('elastic-apm')
+var agent = require('../agent').singleton
```
Comment (watson) [inline lib/instrumentation/shimmer.js]: Why not `require('../../')`?

File: test/parsers.js [hunk 5]
```diff
@@ -381,6 +381,12 @@ test('#parseError()', function (t) {
       _conf: {
         sourceLinesErrorAppFrames: 5,
         sourceLinesErrorLibraryFrames: 5
+      },
+      logger: {
```
Comment (watson) [inline test/parsers.js]: Consider DRY'ing this up (is the DRY ruby'ism known in outside of the Ruby world?)

File: test/queue.js [hunk 6]
```diff
@@ -7,7 +7,13 @@ var Queue = require('../lib/queue')
 test('maxQueueSize', function (t) {
   var opts = {
     maxQueueSize: 5,
-    flushInterval: 1e6
+    flushInterval: 1e6,
+    logger: {
```
Comment (watson) [inline test/queue.js]: Consider DRY'ing this up

File: lib/agent.js (hunk 2 above)
Comment (Qard) [reply lib/agent.js]: I meant for this to just be a possible initialization option. If you have a logger that doesn't initialize this way, you can just initialize it externally and provide the constructed object instead.

File: lib/agent.js (hunk 1 above)
Comment (Qard) [reply lib/agent.js]: fixed

File: lib/parsers.js (hunk 3 above)
Comment (Qard) [reply lib/parsers.js]: fixed

File: lib/instrumentation/shimmer.js (hunk 4 above)
Comment (Qard) [reply lib/instrumentation/shimmer.js]: fixed

File: test/parsers.js (hunk 5 above)
Comment (Qard) [reply test/parsers.js]: fixed

File: test/queue.js (hunk 6 above)
Comment (Qard) [reply test/queue.js]: fixed

File: lib/agent.js (hunk 2 above)
Comment (watson) [reply lib/agent.js]: I'm thinking if a simpler API simply would be to require people to initialize their loggers before hand?

**Comment by Qard** [comment]
I changed it to not have the logger-as-a-function thing, so the level can just be controlled externally. The docs already partially suggested that in the custom logger section, so I just added an extra note to the `logLevel` option.

**Pull request merged**